File: step_13_binary_trees/13.2_medium/IB106_order_of_people_heights/Solution.java
/**
 * Problem: Order of People Heights
 * Difficulty: HARD | XP: 50
 * Source: InterviewBit
 *
 * Given heights[] and infronts[], reconstruct the queue such that
 * person i has exactly infronts[i] people of height >= heights[i] in front.
 * Return the array of heights in the reconstructed order.
 *
 * @author DSA_Nova
 */
import java.util.*;

// ============================================================
// Approach 1: Brute Force (Try all permutations)
// Time: O(n! * n) | Space: O(n)
// ============================================================
class BruteForce {
    static int[] heights;
    static int[] infronts;
    static int[] result;
    static boolean found;

    public static int[] solve(int[] h, int[] inf) {
        heights = h;
        infronts = inf;
        int n = h.length;
        result = new int[n];
        found = false;

        int[] perm = new int[n];
        for (int i = 0; i < n; i++) perm[i] = i;
        permute(perm, 0);
        return result;
    }

    static void permute(int[] perm, int start) {
        if (found) return;
        if (start == perm.length) {
            // Build queue and verify
            int[] queue = new int[perm.length];
            for (int i = 0; i < perm.length; i++) queue[i] = heights[perm[i]];
            for (int pos = 0; pos < perm.length; pos++) {
                int h = heights[perm[pos]];
                int inf = infronts[perm[pos]];
                int count = 0;
                for (int j = 0; j < pos; j++) if (queue[j] >= h) count++;
                if (count != inf) return;
            }
            System.arraycopy(queue, 0, result, 0, queue.length);
            found = true;
            return;
        }
        for (int i = start; i < perm.length; i++) {
            int tmp = perm[start]; perm[start] = perm[i]; perm[i] = tmp;
            permute(perm, start + 1);
            tmp = perm[start]; perm[start] = perm[i]; perm[i] = tmp;
        }
    }
}

// ============================================================
// Approach 2: Optimal (Sort by height desc, insert at index=infronts)
// Time: O(n²) | Space: O(n)
// ============================================================
class Optimal {
    public static int[] solve(int[] heights, int[] infronts) {
        int n = heights.length;
        // Pair (height, infront) and sort by height desc, ties by infront asc
        int[][] people = new int[n][2];
        for (int i = 0; i < n; i++) {
            people[i][0] = heights[i];
            people[i][1] = infronts[i];
        }
        Arrays.sort(people, (a, b) -> a[0] != b[0] ? b[0] - a[0] : a[1] - b[1]);

        LinkedList<Integer> result = new LinkedList<>();
        for (int[] p : people) {
            result.add(p[1], p[0]); // insert height at index=infront
        }

        return result.stream().mapToInt(Integer::intValue).toArray();
    }
}

// ============================================================
// Approach 3: Best (Same greedy — clean production form)
// Time: O(n²) | Space: O(n)
// Note: O(n log n) with BIT/Segment Tree for finding k-th free slot
// ============================================================
class Best {
    public static int[] solve(int[] heights, int[] infronts) {
        int n = heights.length;
        int[][] people = new int[n][2];
        for (int i = 0; i < n; i++) {
            people[i][0] = heights[i];
            people[i][1] = infronts[i];
        }
        // Sort: tallest first, ties broken by fewer-infronts first
        Arrays.sort(people, (a, b) -> a[0] != b[0] ? b[0] - a[0] : a[1] - b[1]);

        // ArrayList for simpler indexed insertion
        ArrayList<Integer> result = new ArrayList<>();
        for (int[] p : people) {
            result.add(p[1], p[0]);
        }

        return result.stream().mapToInt(Integer::intValue).toArray();
    }
}

// Main driver
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Order of People Heights ===\n");

        int[][] heightsArr  = {{5, 3, 2, 6, 1, 4}, {6, 5, 4, 3, 2, 1}, {1, 2}};
        int[][] infrontsArr = {{0, 1, 2, 0, 3, 2}, {0, 0, 0, 0, 0, 0}, {1, 0}};
        int[][] expecteds   = {{5, 3, 2, 1, 6, 4}, {1, 2, 3, 4, 5, 6}, {2, 1}};

        for (int t = 0; t < heightsArr.length; t++) {
            int[] o = Optimal.solve(heightsArr[t], infrontsArr[t]);
            int[] h = Best.solve(heightsArr[t], infrontsArr[t]);
            boolean pass = Arrays.equals(o, expecteds[t]) && Arrays.equals(h, expecteds[t]);

            System.out.println("Heights:  " + Arrays.toString(heightsArr[t]));
            System.out.println("Infronts: " + Arrays.toString(infrontsArr[t]));
            System.out.println("Optimal:  " + Arrays.toString(o));
            System.out.println("Best:     " + Arrays.toString(h));
            System.out.println("Expected: " + Arrays.toString(expecteds[t]));
            System.out.println("Pass:     " + pass);
            System.out.println();
        }
    }
}

File: step_13_binary_trees/13.2_medium/IB106_order_of_people_heights/solution.py
"""
Problem: Order of People Heights
Difficulty: HARD | XP: 50
Source: InterviewBit

You are given two arrays A (heights) and B (infronts).
A[i] is the height of person i. B[i] is the number of people
in front of person i who are taller than or equal to them.
Reconstruct the original queue (the position each person stands in).
Return an array of heights in the reconstructed order.
"""
from typing import List
from itertools import permutations


# ============================================================
# APPROACH 1: BRUTE FORCE (Try all permutations, verify constraints)
# Time: O(n! * n) | Space: O(n)
# ============================================================
def brute_force(heights: List[int], infronts: List[int]) -> List[int]:
    """
    Try every permutation of people. For each permutation, verify that
    every person satisfies their 'infronts' constraint. Return the valid one.
    """
    n = len(heights)
    people = list(range(n))  # indices of people

    for perm in permutations(people):
        queue = [heights[i] for i in perm]
        valid = True
        for pos, idx in enumerate(perm):
            h = heights[idx]
            inf = infronts[idx]
            # Count people in front (positions 0..pos-1) with height >= h
            count = sum(1 for j in range(pos) if queue[j] >= h)
            if count != inf:
                valid = False
                break
        if valid:
            return queue

    return []  # should not reach here on valid input


# ============================================================
# APPROACH 2: OPTIMAL (Sort by height desc, insert at index=infronts)
# Time: O(n²) | Space: O(n)
# ============================================================
def optimal(heights: List[int], infronts: List[int]) -> List[int]:
    """
    Greedy insight: pair each person as (height, infronts) and sort
    by height descending (ties broken by infronts ascending).

    After sorting, insert each person at index `infronts` in the result list.
    Why this works:
    - Process tallest people first. When we insert person (h, k), all people
      already in the result are >= h, so inserting at index k means exactly
      k people taller are in front — satisfying the constraint.
    - Shorter people inserted later don't affect the taller people's counts.
    """
    people = sorted(zip(heights, infronts), key=lambda x: (-x[0], x[1]))
    result = []
    for h, k in people:
        result.insert(k, h)
    return result


# ============================================================
# APPROACH 3: BEST (Same greedy, with linked-list insight for O(n log n))
# Time: O(n²) in Python (list.insert is O(n)); O(n log n) with BIT/seg tree
# Space: O(n)
# ============================================================
def best(heights: List[int], infronts: List[int]) -> List[int]:
    """
    Same greedy strategy as Approach 2. For truly O(n log n) insertion,
    one would use a Binary Indexed Tree (BIT/Fenwick tree) to find the
    k-th free slot in O(log n). In Python, we implement the clean greedy
    version (identical to Approach 2) as the 'best' production approach,
    since list.insert is O(n) and the BIT variant is complex.

    BIT approach summary (for interview discussion):
    - Maintain a BIT over n positions. All positions start as "free" (value=1).
    - Sort by height desc. For each (h, k): find the (k+1)-th free position
      using binary search on BIT prefix sums → O(log n). Mark it used.
    - This gives O(n log n) total.
    """
    people = sorted(zip(heights, infronts), key=lambda x: (-x[0], x[1]))
    result = []
    for h, k in people:
        result.insert(k, h)
    return result


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Order of People Heights ===\n")

    test_cases = [
        (
            [5, 3, 2, 6, 1, 4],
            [0, 1, 2, 0, 3, 2],
            [5, 3, 2, 1, 6, 4],  # one valid arrangement
        ),
        (
            [6, 5, 4, 3, 2, 1],
            [0, 0, 0, 0, 0, 0],
            [1, 2, 3, 4, 5, 6],  # All k=0: each in strictly ascending order so no one taller is ahead
        ),
        (
            [1, 2],
            [1, 0],
            [2, 1],
        ),
    ]

    for heights, infronts, expected in test_cases:
        # Brute force only for small inputs
        b = brute_force(heights[:], infronts[:]) if len(heights) <= 7 else "skipped"
        o = optimal(heights[:], infronts[:])
        h = best(heights[:], infronts[:])

        def verify(result, hts, inf):
            if not result:
                return False
            n = len(result)
            for i in range(n):
                # Find which original person is at position i:
                # since result contains heights, we match by checking infronts property
                pass
            return result == expected  # simplified check against expected

        status = "PASS" if o == expected and h == expected else "FAIL"
        print(f"Heights:  {heights}")
        print(f"Infronts: {infronts}")
        print(f"  Brute:    {b}")
        print(f"  Optimal:  {o}")
        print(f"  Best:     {h}")
        print(f"  Expected: {expected}  [{status}]\n")
